HDDS-12742. RDBStoreAbstractIterator does redundant key and value native calls unnecessarily

swamirishi opened the pull request:
## What changes were proposed in this pull request?
Currently RDBStoreAbstractIterator does redundant key & value native calls multiple times, even though the entry is already present in the heap memory. 

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-12742

## How was this patch tested?
Existing unit tests

szetszwo commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -94,16 +98,25 @@ private void setCurrentEntry() {
   public final boolean hasNext() {
-    return rocksDBIterator.get().isValid() &&
-        (prefix == null || startsWithPrefix(key()));
+    try {
+      if (!seekDoneAtleastOnce) {
+        throw new NoSuchElementException("Cannot check iterator hasNext before seeking either to start of iter or " +
+            "seeking to a specific key");
+      }
+      return currentEntry != null &&
+          (prefix == null || startsWithPrefix(currentEntry.getKey()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
```
We should introduce new exceptions.  Why the exceptions were not needed previously?

swamirishi replied:
key() call from iterator is what is inefficient. In the implementation of this class it would be called from the iterator over and over again which would involve a buffer copy. In byte array it would be direct buffer to indirect buffer copy and in codec buffer it would be direct buffer to direct buffer copy which is still inefficient and unnecessary.

szetszwo reviewed (commented):
@swamirishi , Thanks for working not this!  

> ...  native calls multiple times, ...

I guess the native call is `rocksDBIterator.get().isValid()`?  It is not clear from the description.

The usual way implementing an iterator is to add a `boolean hasNext` field and then set it when moving the iterator; see below:
```diff
diff --git a/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java b/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java
index 75104a55ed..a28693c2b3 100644
--- a/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java
+++ b/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java
@@ -38,6 +38,7 @@ abstract class RDBStoreAbstractIterator<RAW>
   private final ManagedRocksIterator rocksDBIterator;
   private final RDBTable rocksDBTable;
   private Table.KeyValue<RAW, RAW> currentEntry;
+  private boolean hasNext;
   // This is for schemas that use a fixed-length
   // prefix for each key.
   private final RAW prefix;
@@ -47,6 +48,7 @@ abstract class RDBStoreAbstractIterator<RAW>
     this.rocksDBIterator = iterator;
     this.rocksDBTable = table;
     this.prefix = prefix;
+    setHasNext(rocksDBIterator.get().isValid());
   }
 
   /** @return the key for the current entry. */
@@ -85,17 +87,22 @@ public final void forEachRemaining(
   }
 
   private void setCurrentEntry() {
-    if (rocksDBIterator.get().isValid()) {
+    final boolean isValid = rocksDBIterator.get().isValid();
+    if (isValid) {
       currentEntry = getKeyValue();
     } else {
       currentEntry = null;
     }
+    setHasNext(isValid);
+  }
+
+  private void setHasNext(boolean isValid) {
+    hasNext = isValid && (prefix == null || startsWithPrefix(key()));
   }
 
   @Override
   public final boolean hasNext() {
-    return rocksDBIterator.get().isValid() &&
-        (prefix == null || startsWithPrefix(key()));
+    return hasNext;
   }
 
   @Override
```

swamirishi commented:
@szetszwo I have updated the description and changed the scope of the patch here. Please check it out. I am working on benchmarking the proposed changes comparing to our existing implementation. I should have the numbers soon and will publish it here soon.

szetszwo commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreCodecBufferIterator.java:
```diff
@@ -84,38 +92,77 @@ CodecBuffer getFromDb() {
 
-  private final Buffer keyBuffer;
-  private final Buffer valueBuffer;
+  private final Stack<RawKeyValue<Buffer>> availableBufferStack;
+  private final Set<RawKeyValue<Buffer>> inUseBuffers;
+  private final Object bufferLock;
```
`CodecBuffer` already uses a netty buffer pool.  The buffer stack is not needed.

szetszwo commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -78,38 +82,50 @@ final RAW getPrefix() {
-  public final Table.KeyValue<RAW, RAW> next() {
-    setCurrentEntry();
-    if (currentEntry != null) {
+  public final synchronized UncheckedAutoCloseableSupplier<RawKeyValue<RAW>> next() {
```
Why it needs `synchronized`?  Will an iterator object be invoked by multiple threads?

szetszwo replied:
Even for ConcurrentHashMap, 

> ..., iterators are designed to be used by only one thread at a time. ...

See https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

swamirishi replied:
yeah I removed the synchronized method. I will make the next set of PRs into a SplitIterator

szetszwo reviewed (commented):
@swamirishi , thanks for working on this!  Please see the comment/question inlined.

swamirishi commented:
@szetszwo @kerneltime I compared the changes with and without the buffer stack. Allocating direct byte buffer is still a significant portion of the Iteration.
![40_10Benchmark](https://github.com/user-attachments/assets/c3d9c559-2575-4fec-aa41-ef39a565a7a3)
![40_1000Benchmark](https://github.com/user-attachments/assets/55c790ba-48fe-4f82-9cc9-46dd033d0af5)
![400_10Benchmark](https://github.com/user-attachments/assets/2eedcb5b-1b79-4709-8e3a-ac5bb2dfb363)
![400_1000Benchmark](https://github.com/user-attachments/assets/0e2c9bf6-7b75-4989-a411-692b4861cabd)
![4000_10Benchmark](https://github.com/user-attachments/assets/cc7bd160-1fcc-4059-bd0f-6d938a984ade)
![4000_1000Benchmark](https://github.com/user-attachments/assets/d2e4b075-eed9-48de-82c3-20d396e57dca)
![40000_10Benchmark](https://github.com/user-attachments/assets/264c22c8-dda0-4d6f-aa44-36fa28ec2e9b)
![40000_1000Benchmark](https://github.com/user-attachments/assets/1cfb91a8-3dae-4f58-a267-c5a6bb5b2843)

```
package org.apache.hadoop.hdds.utils.db;

import static org.apache.hadoop.ozone.om.OMConfigKeys.OZONE_OM_SNAPSHOT_DB_MAX_OPEN_FILES;

import java.io.IOException;
import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import org.apache.hadoop.hdds.conf.OzoneConfiguration;
import org.apache.hadoop.ozone.om.OMConfigKeys;
import org.apache.hadoop.ozone.om.OmMetadataManagerImpl;
import org.apache.ratis.util.function.CheckedFunction;
import org.apache.ratis.util.function.UncheckedAutoCloseableSupplier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Utility to benchmark splitIterator.
 */
public final class BenchmarkOmDBCodecBufferIter {
  private static final Logger LOG = LoggerFactory.getLogger(BenchmarkOmDBCodecBufferIter.class);
  private BenchmarkOmDBCodecBufferIter() {
  }

  public static void main(String[] args) throws IOException {
    OzoneConfiguration ozoneConfiguration = new OzoneConfiguration();
    ozoneConfiguration.setInt(OZONE_OM_SNAPSHOT_DB_MAX_OPEN_FILES, -1);
    ozoneConfiguration.set(OMConfigKeys.OZONE_OM_DB_DIRS, args[0]);
    RocksDBConfiguration rocksDBConfiguration = ozoneConfiguration.getObject(RocksDBConfiguration.class);
    Long timeToLive = Long.valueOf(args[2]);
    int numberOfBuffers = Integer.valueOf(args[3]);
    long loggingThreshold = Long.parseLong(args[1]);
    ozoneConfiguration.setFromObject(rocksDBConfiguration);
    OmMetadataManagerImpl omMetadataManager = new OmMetadataManagerImpl(ozoneConfiguration, null);
    AtomicLong counter = new AtomicLong();
    long startTime = System.currentTimeMillis();
    AtomicBoolean finished = new AtomicBoolean(true);
    Queue<UncheckedAutoCloseableSupplier<RawKeyValue<CodecBuffer>>> q = new LinkedList<>();
    CheckedFunction<UncheckedAutoCloseableSupplier<RawKeyValue<CodecBuffer>>, Boolean, IOException> func = kv -> {
      long id = counter.incrementAndGet();
      q.add(kv);
      if (id % loggingThreshold == 0) {
        long time = System.currentTimeMillis();
        LOG.info("Benchmark: " + id + "\t" + (time - startTime) + "\t" + StringCodec.get().fromCodecBuffer(kv.get().getKey()));
        finished.set(time - startTime <= timeToLive);
      }
      if (q.size() > numberOfBuffers) {
        q.poll().close();
      }
      return finished.get();
    };
    RDBStore rdbStore = (RDBStore) omMetadataManager.getStore();
    RDBTable rdbTable = rdbStore.getTable("fileTable");
    RocksDatabase rocksDatabase = rdbStore.getDb();
    try (TableIterator<CodecBuffer, UncheckedAutoCloseableSupplier<RawKeyValue<CodecBuffer>>> itr =
             new RDBStoreCodecBufferIterator(rocksDatabase.newIterator(rdbTable.getColumnFamily()), rdbTable,
                 null, numberOfBuffers + 2)) {
      while (itr.hasNext()) {
        if (!func.apply(itr.next())) {
          break;
        }
      }
      while (!q.isEmpty()) {
        q.poll().close();
      }
    }
    omMetadataManager.getStore().close();
  }
}


```

szetszwo commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -78,32 +82,44 @@ final RAW getPrefix() {
-  public final Table.KeyValue<RAW, RAW> next() {
-    setCurrentEntry();
-    if (currentEntry != null) {
+  public final UncheckedAutoCloseableSupplier<RawKeyValue<RAW>> next() {
```
@swamirishi ,  The current single-threaded API is below
```java
    TypedTable<ContainerID, String> table = ...;
    try (TableIterator<ContainerID, ? extends Table.KeyValue<ContainerID, String>> i = table.iterator()) {
      final KeyValue<ContainerID, String> keyValue = i.next();
      // process it
    }
```
What is the multi-threaded API in your mind?  What would the above code look like?

swamirishi replied:
ThreadLocal will not work. If you see the change in the algorithm, the value set here in this is for the next value in the iterator, the thread receives the prev value that is stored in the memory.

swamirishi replied:
```
TableIterator<UncheckAutoCloseableSupplier<RawKeyValue>> itr = ...;
boolean tryAdvance(Consumer<? super T> action) {
        lock.lock();
        final UncheckAutoCloseableSupplier<RawKeyValue> keyValue;
        try {
          if (itr.hasNext()) {

             keValue = itr.next();
          }


       } finally {
         lock.unlock();
      }
     try {
        // DeserialedKeyValue  = deserialize(keyValue);
       
     }  finally {
       keyValue.close();
     } 
    //process DeserialedKeyValue     
}

```
Above is the high level idea I have for the parallel iterator.

szetszwo replied:
> TableIterator<UncheckAutoCloseableSupplier<RawKeyValue>> itr = ...;

What is "..."?  How to obtain an iterator?  What are the parameters?

> boolean tryAdvance(Consumer<? super T> action) { ....

I guess you are not saying that all OM, recon code have to implement a SplitIterator.

My question is the API using the new iterator -- how the OM, Recon code will look like?

swamirishi replied:
Table interface will have splitIterator function.
```

public class RDBTable {
  // This function would be part of the table interface.
 public void performOperationInParallel(byte[] prefix, Consumer<RawKeyValue<byte[]>> op) {
     TableIterator<UncheckAutoCloseableSupplier<RawKeyValue>> itr = ...;
     SplitTableIterator itr = new SplitTableIterator(itr);
     performOperationInParallel(itr, op);
  }

  public void performOperationInParallel(CodecBuffer prefix, Consumer<RawKeyValue<CodecBuffer>> op) {
     TableIterator<UncheckAutoCloseableSupplier<RawKeyValue>> itr = ...;
     SplitTableIterator itr = new SplitTableIterator(itr);
     performOperationInParallel(itr, op)
  }

  private performOperationInParallel(SplitTableIterator splittableIterator, Consumer<RawKeyValue<CodecBuffer>> op) {
   ThreadPoolExecutor pool = ...;
   pool.submit(splittableIterator.tryAdvance()); 
   .....
}
}
```
Just chalking out the high level idea of the interface changes in the next set of PRs

szetszwo replied:
How about simply passing an executor and a boolean `ordered` to `table.iterator(..)`?
```java
    TypedTable<ContainerID, String> table = ...;
    Executor executor = ...;
    boolean ordered = true|false; // Are the returned elements preserving the original iteration order?
    try (TableIterator<ContainerID, ? extends Table.KeyValue<ContainerID, String>> I
        = table.iterator(executor, orderd)) {
      final KeyValue<ContainerID, String> keyValue = i.next();
      // process it as before
    }
```
The OM, Recon code should not care about the implemetation details such as how the buffers gets released.

swamirishi replied:
Yeah the typed table interface will translate the consumer operation

szetszwo commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreCodecBufferIterator.java:
```diff
@@ -84,38 +92,77 @@ CodecBuffer getFromDb() {
 
-  private final Buffer keyBuffer;
-  private final Buffer valueBuffer;
+  private final Stack<RawKeyValue<Buffer>> availableBufferStack;
```
Stack is not the best data structure in this case.  We may use a  [BlockingQueue](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html).

It is even better to use  `ThreadLoacl<Buffer>` as mentioned my previous comment.

swamirishi replied:
I initially used a queue but I realized we may not have to allocate all buffers all the time. LIFO is better than FIFO in such cases. My buffer allocation would be minimal.

swamirishi replied:
Just because we have initialized a Buffer object it doesn't mean we have allocated anything yet. It will still do a lazy allocation. Use of stack will minimize if there are not many concurrent threads in lock code block than the number of max buffers the allocation that is why it is better to use a stack. Moreover stack.pop() is a very cheap operation. There would be only one thread waiting for a non empty stack which should be fine. There should be no contention as well in this case ideally. We need the synchronized block only for the published subscriber model, optmize on the wait() & notify()

swamirishi replied:
Instead of having a custom blocking stack implementation I have changed it to using BlockingDeque instead which provides these functionalities. As part of this change I had to handle race conditions with the inuseBufferSet values. Since we need to handle 2 data structures in a consistently. 
Please take a look at this and check which is lesser complex we can go ahead with whatever approach we all align with.

szetszwo replied:
> ... LIFO is better than FIFO in such cases. ...

Why LIFO is better than FIFO? 

BTW, for LIFO, we have https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingDeque.html

> ... My buffer allocation would be minimal with a stack

For any data structures, the allocation can be minimal since we can have lazy allocation -- allocate only when it is used.

> ... We need the synchronized block only for the published subscriber model, optmize on the wait() & notify()

It also makes the code complicated.  Prone to a deadlock.

szetszwo reviewed (commented):
@swamirishi , thanks for the update!  Please see the comments inlined

szetszwo commented:
> ThreadLocal will not work.

It will work.

> Moreover ThreadLocal will end up creating too many buffers which IMO is very wasteful.

No, it won't.

@swamirishi , Since this is a new iterator implementation, let's create a new class for it.  It is quite invasive and risky to change the existing code.  We should make it configurable.  If we find some bugs later on, we could set the conf to use the old iterator.

swamirishi commented:
> > ThreadLocal will not work.
> 
> It will work.
> 
> > Moreover ThreadLocal will end up creating too many buffers which IMO is very wasteful.
> 
> No, it won't.
> 
> @swamirishi , Since this is a new iterator implementation, let's create a new class for it. It is quite invasive and risky to change the existing code. We should make it configurable. If we find some bugs later on, we could set the conf to use the old iterator.

@szetszwo I will go ahead and make this as a separate implementation. But the downstream usages of these classes would be really messy with if else conditions the idea of which I don't like. I would be still ok if we eventually want to get rid of the old implementation.

sumitagrawl reviewed (commented):
@swamirishi 
1. Currently RDBStoreCodecBufferIterator is wrapped by RawIterator in TypedTable and its thread safe as per use case.
2. For Split use case or having deserialization separate, we can have another iterator implementation like RDBStoreCodecBuffer`Split`Iterator

So in this way, we need not impact existing implementation by wrapping with UncheckedAutoCloseableSupplier OR another iterator wrapper for access.

And CodecBufferIterator can be simple with just one buffer as default implementation.

swamirishi commented:
@sumitagrawl As discussed offline I have added a split iterator implementation which would show how the codecbufferIterator would be used.

kerneltime reviewed (commented):
Finished one round of looking into the code. This is rather critical piece of code where error handling will be critical. Will spend some more time looking into the tests.

sumitagrawl commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:
```diff
@@ -218,14 +219,63 @@ public TableIterator<byte[], KeyValue<byte[], byte[]>> iterator()
+    return iterator(prefix, 2);
+  }
+
+  TableIterator<CodecBuffer, UncheckedAutoCloseableSupplier<RawKeyValue<CodecBuffer>>> iterator(
```
We can remove this iterator creation here and be part of specialized method, as this needs to be wrapped for handling Buffer release. No need expose.

sumitagrawl reviewed (commented):
@swamirishi We need some code re-organization to limit the impact of CodecBufferIterator to be only limited to that class.

sumitagrawl commented:
@swamirishi Can you share test code run to capture performance for split iterator with normal iterator

sumitagrawl commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -150,5 +181,21 @@ public final void removeFromDB() throws IOException {
+    releaseEntry();
+  }
+
+  public static final class AutoCloseableRawKeyValue<RAW> extends RawKeyValue<RAW> implements AutoCloseable {
```
Can rename class as CloseableRawKeyValue, can remove Auto keyword

swamirishi replied:
Closeable is for specifically throwing IOExceptions. AutoCloseable is the base interface. Since we are not throwing any exceptions we should call it AutoCloseableRawKeyValue ideally

sumitagrawl commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -150,5 +181,21 @@ public final void removeFromDB() throws IOException {
+  public static final class AutoCloseableRawKeyValue<RAW> extends RawKeyValue<RAW> implements AutoCloseable {
+    private final UncheckedAutoCloseableSupplier<RawKeyValue<RAW>> keyValue;
+
+    private AutoCloseableRawKeyValue(ReferenceCountedObject<RawKeyValue<RAW>> kv) {
```
This itself can be RefrenceCounted AutoClosable class, instead of wrapping at caller.
- retain() --> add reference count
- release()/close() --> release reference count and when "0", close the object.

swamirishi replied:
No we should only return the autoCloseable value and not the reference counted object itself if that is what you mean

sumitagrawl commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -125,23 +151,28 @@ public final void seekToLast() {
-    if (currentEntry != null) {
-      delete(currentEntry.getKey());
+    if (previousKeyValue != null) {
+      delete(previousKeyValue.getKey());
```
since being reference counted, if after release, this is called can have issue. Need keep reference Counted only, and throw exception if its closed

swamirishi replied:
RemoveFromDB should not be called after release. That is the whole point of returning a CloseableValue. Even getKey() and getValue() won't be valid.

sumitagrawl commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -78,32 +84,52 @@ final RAW getPrefix() {
+    if (hasNext()) {
+      AutoCloseableRawKeyValue<RAW> entry = new AutoCloseableRawKeyValue<>(currentEntry);
+      this.previousKeyValue = currentEntry.get();
       rocksDBIterator.get().next();
```
rocksDbIterator.get().next() can also be moved to setCurrentEntry(), as we are getting next entry, updating hasNext flag together.

swamirishi replied:
no we cannot do that. What about for the first entry. You shouldn't call next for the first entry in the iterator

sumitagrawl commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/Table.java:
```diff
@@ -420,4 +420,8 @@ public int hashCode() {
   }
+
+  /** A {@link org.apache.hadoop.hdds.utils.db.TableSpliterator} to split iterate {@link KeyValue}s. */
+  interface KeyValueSpliterator<KEY, VALUE> extends TableSpliterator<KEY, KeyValue<KEY, VALUE>> {
```
Do we need TableSpliterator defined?

swamirishi replied:
I removed it.

sumitagrawl reviewed (commented):
@swamirishi above is much better, ReferenceCounted and AutoClosable should be single Class type, no need have 2 wrap. This will further simplify code, and few comments.

swamirishi commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RawSpliterator.java:
```diff
@@ -0,0 +1,128 @@
+    TableIterator<RAW, UncheckedAutoCloseableSupplier<RawKeyValue<RAW>>> itr = getRawIterator(prefix,
+        startKey, maxParallelism);
+    this.lock = new ReentrantLock();
+    this.maxNumberOfAdditionalSplits = new AtomicInteger(maxParallelism - 1);
```
No we need it. When we pass this to StreamSupport.stream(spliterator) all splits are supposed to be ideally acting independently. One split should not impact another split. That is why we are referenceCounting here.

swamirishi commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java:
```diff
@@ -30,30 +32,34 @@
   // This is for schemas that use a fixed-length
   // prefix for each key.
   private final RAW prefix;
+  private Boolean hasNext;
```
Iterators are not thread safe in general. So this is not supposed to be used by multiple threads.

szetszwo commented:
> ... But the downstream usages of these classes would be really messy ...

@swamirishi , No, it won't be.  We could have a method to read the conf and take care which implementation of the iterator to use.  If the downstream usages is messy, the API definitely is not well-designed.

swamirishi commented:
> I strongly oppose and -1 on changing the existing iterator for the new implementation since this is an invasive and risky change. We need the old iterator in case that we find bugs later.

@szetszwo I have added the change behind a flag please take a look at the change if the changes are good.

szetszwo commented:
@sumitagrawl , @swamirishi , 

Below are the current API methods
```java
//Table.java
  TableIterator<KEY, ? extends KeyValue<KEY, VALUE>> iterator() throws IOException;
  TableIterator<KEY, ? extends KeyValue<KEY, VALUE>> iterator(KEY prefix) throws IOException;
```
The new API method should look like:
```java
  TableIterator<KEY, ? extends KeyValue<KEY, VALUE>> iterator(KEY prefix, Executor executor, boolean ordered) throws IOException;
```
See also [this comment](https://github.com/apache/ozone/pull/8203#discussion_r2049466909).  The existing methods could should just pass null for executor and true for ordered to the new method.  When executor is non-null, the new implementation will use it for parallel execution.  When it is null, it will use the existing (single-thread) implementation.

`SplitIterator` and reference count are all implementation detail in the new implementation.  It should not be exposed to the API.  (Just like that iterator can use `CodecBuffer` or `byte[]`.  They are not in the API.)

It there is a need, we could add a nextAsync() method to TableIterator for asynchronous execution at the caller side, i.e. the caller can process the elements in parallel.
```java 
//TableIterator
  CompletableFuture<T> nextAsync();
```

swamirishi commented:
I am working on breaking this PR into smaller components as suggested by @szetszwo I have created https://issues.apache.org/jira/browse/HDDS-13151 which would be just have the refactoring changes for iterators in the code base.

szetszwo commented:
> ... Any implementation of Spliterator would give the downstream api users the flexibility to use the object in the way they deem fit may be wanting to submit the spliterator to a forkjoin pool to do a parallel iteration using multiple threads. ...

The point here is to minimize the change to the existing code.  The new parallel iterator currently is an experimental feature.  Once the parallel iterator has become stable in the future, we may add `Spliterator` to the API if there is a need.

swamirishi commented:
> > ... Any implementation of Spliterator would give the downstream api users the flexibility to use the object in the way they deem fit may be wanting to submit the spliterator to a forkjoin pool to do a parallel iteration using multiple threads. ...
> 
> The point here is to minimize the change to the existing code. The new parallel iterator currently is an experimental feature. Once the parallel iterator has become stable in the future, we may add `Spliterator` to the API if there is a need.

Why can't we wrap the iterator inside a spliterator and then return it. Even if the api tries to use it as a spliterator it would still be an iterator inside a synchronized spliterator. Something like this, when the flag is disabled we should be just returning a SynchronizedSpliterator which doesn't have much logic here. Do you see any risk if we do this @szetszwo 

```
import java.util.Iterator;
import java.util.Spliterator;
import java.util.function.Consumer;

public class SynchronizedSpliterator<T> implements Spliterator<T> {

    private final Iterator<T> iterator;

    public SynchronizedSpliterator(Iterator<T> iterator) {
        this.iterator = iterator;
    }

    @Override
    public synchronized boolean tryAdvance(Consumer<? super T> action) {
        if (iterator.hasNext()) {
            action.accept(iterator.next());
            return true;
        }
        return false;
    }

    @Override
    public Spliterator<T> trySplit() {
        return null; // No splitting support
    }

    @Override
    public long estimateSize() {
        return Long.MAX_VALUE; // Unknown size
    }

    @Override
    public int characteristics() {
        return ORDERED | NONNULL; // Customize as needed
    }
}

```

szetszwo commented:
> ...  Do you see any risk if we do this ...

@swamirishi , yes.  

```java
     public synchronized boolean tryAdvance(Consumer<? super T> action) {
```
Are you sure that it is correct to have `synchronized` in `tryAdvance`?  The Java subclass implementations do not have `synchronized`.

Moreover, the change looks simple today but it may require more changes tomorrow (Just like this PR, it keeps growing unexpectedly.)  Let's  have a solid implementation before making a bigger change to the existing code.  `Spliterator` is just for convenience but not any real improvement.  Of course, the new implementation may still use it internally.

layavadi commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:
```diff
@@ -37,14 +40,15 @@
  * metadata store content. All other user's using Table should use TypedTable.
  */
 @InterfaceAudience.Private
```
Is this is still Private

szetszwo commented:
The [SynchronizedSpliterator](https://github.com/apache/ozone/pull/8203#issuecomment-2932317879) approach is inefficient -- Spliterator is designed to split a collection into multiple parts and then iterate the individual parts in parallel.

For example, suppose there is a list with 10,000,000 elements and it is going to be processed by 100 threads.
- Spliterator is designed to do the split first, say splitting it into 1000 parts with 10,000 elements each.  Then, the 100 threads only have to be synchronized 1000 times for picking up the individual parts.
- For SynchronizedSpliterator, since it has synchronized at tryAdvance(..), the 100 threads have to be synchronized 10,000,000 times for picking up the individual elements.
